blog/2025-03-25-K3s-bootstrap.md

```diff
@@ -0,0 +1,120 @@
+drwxr-xr-x 4 root root 4.0K Mar 19 06:22 ..
+drwxr-xr-x 4 root root 4.0K Mar 19 06:28 82142f5157c67effc219aeefe0bc03e0460fc62b9fbae9e901270c86b5635d53
+lrwxrwxrwx 1 root root   90 Mar 19 06:28 previous -> /var/lib/rancher/k3s/data/82142f5157c67effc219aeefe0bc03e0460fc62b9fbae9e901270c86b5635d53
+drwxr-xr-x 4 root root 4.0K Mar 19 06:30 b13851fe661ab93938fc9a881cdce529da8c6b9b310b2440ef01a860f8b9c3a9
+lrwxrwxrwx 1 root root   90 Mar 19 06:30 current -> /var/lib/rancher/k3s/data/b13851fe661ab93938fc9a881cdce529da8c6b9b310b2440ef01a860f8b9c3a9
+drwxr-xr-x 4 root root 4.0K Mar 19 06:40 .
+```
+
+Additionally, K3s includes embedded Helm charts for deploying critical services such as CoreDNS, Traefik, and local storage. These embedded charts, formatted as yaml files, can be found in the control plane nodes in the directory: `/var/lib/rancher/k3s/server/manifests`.
+
+For 67MB our K3s binary includes a lot of stuff!
+
+
+## The Boot Sequence, Step-by-Step 👣
+Now that we established how K3s is carrying its embedded tools, we can explore the boot up sequence. Let us look at the typical logs you can find in `journalctl` when deploying a control-plane or K3s server instance. It all starts with:
+
+```
+Starting Lightweight Kubernetes...
+```
+
+And then:
+
+```
+/usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
+```
+
+This part is checking for a network manager utility which must be disabled as described in the docs. It configures some parts of the network stack, specifically the routing tables, which conflict with Kubernetes networking, and that is why we verify if it was correctly disabled.
+
+The next message should look familiar
+
+```
+Acquiring lock file /var/lib/rancher/k3s/data/.lock
+Preparing data dir /var/lib/rancher/k3s/data/f8e9b5e7d85085972f4a9ddfd539d4dcf887be2e380a55f415c93cac5516dad5
+```
+
+When this message is shown, the directory where K3s deploys the embedded binaries has already been created. At this point, K3s will extract the binaries. We use the lock to avoid concurrent modifications, preventing K3s embedded commands like `kubectl` or `ctr` from executing on this in-process bootstrap.
+
+The next block of logs point at the K3s version and the datastore. In this case, I am deploying the default K3s which means sqlite. For more information on the different datastores available check this [link](https://docs.k3s.io/datastore)
+
+```
+Starting k3s v1.32.2+k3s1
+Configuring sqlite3 database connection pooling: maxIdleConns=2, maxOpenConns=0, connMaxLifetime=0s
+Configuring database table schema and indexes, this may take a moment...
+Database tables and indexes are up to date
+Kine available at unix://kine.sock
+```
+
+Once the datastore is available k3s locks the bootstrap key. This key contains essential data for other control-plane members when using HA mode. We need to lock it so that there is only one member adding that information. We are not using HA mode, so this is not relevant in this bootstrap example.
```

Comment:
It is "locked" by creating the bootstrap key in etcd with the hash of the current server token, but with empty data. Since it exists, other cluster members will not start the process of generating new CA certs - they will wait for it to be populated with data. This is done to fix an issue where multiple nodes that started at the same time would both find the key missing and create different cluster CAs.

Reply:
Thanks Brad! The process is complex and I am missing details. I added a small summary :)